### PR TITLE
(GH-38) Include Admin Tag when using Admin Helpers

### LINKED ISSUE #38: Guideline: Admin tag when using Admin helper
Will become required later (put the date into suggestion)

https://github.com/chocolatey/choco/wiki/HelpersReference

### PR DESCRIPTION
**Pull request by gep13**
Closes #38

### CODE REVIEW

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace chocolatey.package.validator.infrastructure.app.rules
+{
+    using System.IO;
+    using System.Linq;
+    using infrastructure.rules;
+    using NuGet;
+
+    public class AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline : BasePackageRule
+    {
+        public override string ValidationFailureMessage { get { return "When Admin Helper Functions are used, it is recommended to include an admin Tag in the nuspec file.  Please see [here](https://github.com/chocolatey/package-validator/wiki/AdminTagShouldBeUsedWhenUsingAdminHelper) for further information and guidance."; } }
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: This is one that is going to have a specific date where it is required. We should explain that in this message.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+            foreach (var packageFile in files)
+            {
+                string extension = Path.GetExtension(packageFile.Path).to_lower();
+                if (extension != ".ps1" && extension != ".psm1") continue;
+
+                var contents = packageFile.GetStream().ReadToEnd().to_lower();
+
+                if (contents.Contains("install-chocolateypackage") || contents.Contains("start-chocolateyprocessasadmin") || contents.Contains("install-chocolateyinstallpackage")
+                    || contents.Contains("install-chocolateypath") || contents.Contains("install-chocolateyenvironmentvariable") || contents.Contains("install-chocolateyexplorermenuitem")
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Install-ChocolateyPath is only that way when using machine

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    || contents.Contains("install-chocolateyfileassociation") || contents.Contains("update-sessionenvironment"))
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Update-SessionEnvironment?

Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: It was on the list in the helper function section of the wiki :-)

Sent from my Windows Phone
________________________________

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 2 above)
Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: @ferventcoder should I take that one out just now then?

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 3 above)
Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Ok, I will get that one removed.  Do you want me to update the wiki article as well, or have you done that already?

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 2 above)
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Maybe? Maybe for now, yes.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 3 above)
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Wiki article will need updated.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 2 above)
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Also would you mind dropping a line for each of these items?

**Comment by gep13** [comment]
@ferventcoder wiki article updated, and pushed your suggestions.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 1 above)
Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: @ferventcoder do you know what that date is going to be yet?

Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Let's target Feb 1.

**Comment by ferventcoder** [comment]
The date I think we'll move it to required is Feb 1.

**Comment by ferventcoder** [comment]
@gep13 if you want to fix up and add a date to the notice, that would be great.

**Comment by gep13** [comment]
@ferventcoder done.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                var contents = packageFile.GetStream().ReadToEnd().to_lower();
+
+                if (contents.Contains("install-chocolateypackage") ||
+                    contents.Contains("start-chocolateyprocessasadmin") ||
+                    contents.Contains("install-chocolateyinstallpackage") ||
+                    contents.Contains("install-chocolateyenvironmentvariable") ||
+                    contents.Contains("install-chocolateyexplorermenuitem") ||
+                    contents.Contains("install-chocolateyfileassociation"))
+                {
+                    return package.Tags.Split(' ').Any(tag => tag.ToLower() == "admin");
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: Now I'm lost, we are returning valid at the end of this, yet this is returning early.

Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: @ferventcoder I "think" this is ok though, right?  I have just added some tests to this guideline, which suggest that it is working as I expect it to.  Thoughts?

Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: It's more that I usually do an early return for catching quick issues, but if I create a variable to return, I usually assign to the variable the output and let it return.

File: src/chocolatey.package.validator.tests/infrastructure.app/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuidelineSpecs.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+        private Stream generate_stream_from_string(string s)
+        {
+            var stream = new MemoryStream();
+            var writer = new StreamWriter(stream);
+            writer.Write(s);
+            writer.Flush();
+            stream.Position = 0;
+            return stream;
+        }
```
Comment (ferventcoder) [inline src/chocolatey.package.validator.tests/infrastructure.app/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuidelineSpecs.cs]: This helper is good enough to move up somewhere for the other specs.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs (hunk 4 above)
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: I mean, there is nothing technically wrong with this, it's more of a code smell to me.

**Comment by gep13** [comment]
@ferventcoder ok, I have fixed up the issues that you mentioned, and also move the method to extract a stream from a string to an extension method called to_stream.

Let me know what you think.

File: src/chocolatey.package.validator/StringExtensions.cs [hunk 6]
```diff
@@ -87,6 +88,16 @@ public static string to_string(this string input)
...
+        public static Stream to_stream(this string input)
+        {
+            var stream = new MemoryStream();
+            var writer = new StreamWriter(stream);
+            writer.Write(input);
+            writer.Flush();
+            stream.Position = 0;
+            return stream;
+        }
+
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/StringExtensions.cs]: Weird that you didn't create this in the .tests project?

Comment (gep13) [inline src/chocolatey.package.validator/StringExtensions.cs]: I thought it made more sense to live with the existing Extensions for String object.  Happy to leave only in Tests project if you feel that is better.

Comment (ferventcoder) [inline src/chocolatey.package.validator/StringExtensions.cs]: No worries.

File: src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                var contents = packageFile.GetStream().ReadToEnd().to_lower();
+
+                if (!(contents.Contains("install-chocolateypackage") ||
+                    contents.Contains("start-chocolateyprocessasadmin") ||
+                    contents.Contains("install-chocolateyinstallpackage") ||
+                    contents.Contains("install-chocolateyenvironmentvariable") ||
+                    contents.Contains("install-chocolateyexplorermenuitem") ||
+                    contents.Contains("install-chocolateyfileassociation")))
+                {
+                    return true;
```
Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: ?

Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: I think this is messed up a bit now.

Comment (gep13) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: In which case I am lost :smile: 

What would you do in this situation?

Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: I wouldn't return early by now. haha, just do the additional check and set valid true or false.

Comment (ferventcoder) [inline src/chocolatey.package.validator/infrastructure.app/rules/AdminTagShouldBeIncludedWhenUsingAdminHelpersGuideline.cs]: No worries, I will make the change. :)

**Pull request merged**